src/storage-client/src/types/connections.rs

```diff
@@ -1029,3 +1056,200 @@ impl SshTunnel {
         Ok((local_port, Box::new(tx)))
     }
 }
+
+use rdkafka::client::{BrokerAddr, NativeClient, OAuthToken};
+use rdkafka::config::RDKafkaLogLevel;
+use rdkafka::consumer::{ConsumerContext, Rebalance};
+use rdkafka::error::{KafkaError, KafkaResult};
+use rdkafka::producer::{DeliveryResult, ProducerContext};
+use rdkafka::types::RDKafkaRespErr;
+use rdkafka::util::Timeout;
+use rdkafka::{Statistics, TopicPartitionList};
+
+/// A client context that supports rewriting broker addresses,
+/// and is also able to spin up ssh tunnels dynamically.
+pub struct DynamicBrokerRewritingClientContext<C> {
+    pub inner: BrokerRewritingClientContext<C>,
+    default_override: Option<SshTunnel>,
+    secrets_reader: Arc<dyn SecretsReader>,
+    error: Mutex<Option<String>>,
+    runtime: Handle,
+}
+
+impl<C> DynamicBrokerRewritingClientContext<C> {
+    /// Constructs a new context that wraps `inner`.
+    pub fn new(
+        inner: C,
+        secrets_reader: Arc<dyn SecretsReader>,
+        runtime: Handle,
+    ) -> DynamicBrokerRewritingClientContext<C> {
+        DynamicBrokerRewritingClientContext {
+            inner: BrokerRewritingClientContext::new(inner),
+            default_override: None,
+            secrets_reader,
+            error: Mutex::new(None),
+            runtime,
+        }
+    }
+
+    /// Adds the default broker rewrite rule.
+    ///
+    /// Connections to brokers that aren't specified in other rewrites will be rewritten to connect to
+    /// `rewrite_host` and `rewrite_port` instead.
+    pub fn add_default_broker_rewrite(&mut self, tunnel: SshTunnel) {
+        self.default_override = Some(tunnel);
+    }
+
+    /// Returns a reference to the wrapped context.
+    pub fn inner(&self) -> &C {
+        self.inner.inner()
+    }
+
+    /// Returns any errors that occurred when setting up
+    /// ssh tunnels.
+    pub fn dynamic_error(&self) -> Option<String> {
+        self.error.lock().expect("poisoned").clone()
+    }
+}
+
+impl<C> ClientContext for DynamicBrokerRewritingClientContext<C>
+where
+    C: ClientContext,
+{
+    const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
+
+    fn rewrite_broker_addr(&self, addr: BrokerAddr) -> BrokerAddr {
+        if self.inner.overridden(&addr) {
+            return self.inner.rewrite_broker_addr(addr);
+        }
+
+        match &self.default_override {
+            Some(default_tunnel) => {
+                let broker = addr.clone();
+                let default_tunnel = default_tunnel.clone();
+                let sr = Arc::clone(&self.secrets_reader);
+                let runtime = self.runtime.clone();
+                // We spawn an additional thread here, and immediately
+                // join it so that the rdkafka thread calling this callback
+                // spends all its time parked, instead of being
+                // continuously woken up when we `block_on` the
+                // ssh tunnel creation.
```

Comment:
I'm confused about this. Why does calling `runtime.block_on` from the rdkafka thread cause wakeups?